evm_loader/program/src/account/state.rs

```diff
@@ -98,37 +103,89 @@ impl<'a> StateAccount<'a> {
             tag => return Err(Error::AccountInvalidTag(*info.key, tag)),
         };
 
-        let revisions = accounts
+        // Initialize the heap before any allocations.
+        Self::init_heap(&info)?;
+
+        // todo: get revision from account
+        let revisions_it = accounts
             .into_iter()
             .map(|account| {
                 let revision = revision(program_id, account).unwrap_or(0);
                 (*account.key, revision)
-            })
-            .collect();
+            });
 
-        let data = Data {
+        // Construct TreeMap (based on the AccountAllocator) from constructed revisions_it. 
+        let mut revisions = TreeMap::<Pubkey, u32>::new();
+        for (key, rev) in revisions_it {
+            revisions.insert(key, &rev);
+        }
+
+        let data = boxx(Data {
             owner,
             transaction,
             origin,
             revisions,
             gas_used: U256::ZERO,
-        };
+        });
 
         super::set_tag(program_id, &info, TAG_STATE, Header::VERSION)?;
         {
             // Set header
             let mut header = super::header_mut::<Header>(&info);
-            header.evm_state_len = 0;
-            header.evm_machine_len = 0;
-            header.data_len = 0;
+            header.evm_state_offset = 0;
+            header.evm_machine_offset = 0;
+
+            let account_data_ptr = info.try_borrow_data()?.as_ptr();
+            let data_obj_addr = (&*data) as *const Data as *const u8;
+            header.data_offset = unsafe {data_obj_addr.offset_from(account_data_ptr) as usize};
         }
 
         Ok(Self {
             account: info,
-            data,
+            data: ManuallyDrop::new(Boxx::into_inner(data)),
         })
     }
 
+    pub fn alloc<'r>(&self, account_storage: &'r ProgramAccountStorage<'a>) -> Result<()> {
```

Comment:
This method and all newly added ones need `#[cfg(target_os = "solana")]`. The compilation errors should guide the usage of `#[cfg(target_os = "solana")]`.